Define processes for enhancement proposals (#4818)
apache/openwhisk: proposals/POEM-1-proposal-for-openwhisk-enhancements.md

<!--
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
-->

# Title
Process for introducing an OpenWhisk Enhancement (POEM)

## Status
* Current state: Draft
* Author: @style95

## Summary

> Introduce a process for suggesting and implementing a substantive OpenWhisk Enhancement (POEM).

Reviewer: ```suggestion
Introduce and document a process for suggesting and implementing a substantive OpenWhisk Enhancement (POEM).
```

A developer or group of developers working together to propose and implement a major new feature or functionality, a new subsystem, or a breaking change should follow the process described herein and open a proposal for consideration under this directory.

## Motivation

As the project grows, more and more issues are getting complex and require multiple parties and an extended period of time to develop them.
We can incubate, manage, and elaborate new ideas in OpenWhisk with a standard way and a well-tracked artifact.

> It would enhance the discoverability of proposals and help community members to involve in the project better.

Reviewer: ```suggestion
The goals are to enhance the discoverability of proposals, and to help community members who want to get involved in the project.
```

## Proposed changes

### Procedures
1. Create a pull request to describe your proposal with [this template](./POEM-N-template.md). The initial state of a proposal should be _"Draft"_.

> 2. [Create a corresponding issue]((/issues/new?template=proposal.md)) to propose a new change based on [this template](../github/ISSUE_TEMPLATE/proposal.md). It is mainly used to track discussion history.

Reviewer: I think the link to issues to be adjusted:
```suggestion
2. [Create a corresponding issue]((../../../issues/new?template=proposal.md)) to propose a new change based on [this template](../github/ISSUE_TEMPLATE/proposal.md). It is mainly used to track discussion history.
```
```suggestion
2. [Create a corresponding issue]((/issues/new?template=proposal.md)) to propose a new change based on [this template](../github/ISSUE_TEMPLATE/proposal.md). It is mainly used to track discussion history.
```
```suggestion
2. [Create a corresponding issue]((/issues/new?template=proposal.md)) to propose a new change based on [this template](../github/ISSUE_TEMPLATE/proposal.md). It is mainly used to track discussion history.
```

> 3. Discuss the proposals with any channels(issue comments, mailing list, Slack, etc) to form a consensus and update your proposal based on comments as much as needed.

Reviewer: ```suggestion
3. Discuss the proposals using to form a consensus and update your proposal based on comments as needed. It is important to be inclusive, and to notify the OpenWhisk community of meaningful changes using the Apache [`dev` list](https://openwhisk.apache.org/community.html) for this project. Other forms of communication such as Slack are OK but any meaningful results should be documented in issues and the `dev` list.
```

> 4. When members form a rough consensus for the proposal. The proposal owner can request a vote via the dev mailing list.

Reviewer: is this vote intended as a predicate to merging a PR for the _proposal_?

> 5. The voting process follows the [Apache Voting guideline](https://www.apache.org/foundation/voting.html). The PR can be merged with the _"In-progress"_ state if the voting is successfully closed without any veto.

Reviewer: Votes may pass by silent assent. Do we want to allow silent assent or require at least 3 +1s to show buy in?
cc @dgrove-oss for input.

Reviewer 2: Not sure how best to formalize this, but I'd hope there would be at least some discussion and positive comments.  The pattern to avoid is a proposal being under-discussed and then merged by silent consent.

> 6. The implementation begins as the proposal is filed into the repo and any volunteer can join the implementation ideally.

Reviewer: Perhaps add that this project adheres to a strict code of conduct that encourages inclusiveness and link to our code of conduct https://www.apache.org/foundation/policies/conduct.html.

7. If the proposal is not accepted or no consensus is formed, the PR is merged with the state, _"Abandoned"_.

> 8. When implementation is included in the master branch, it is supposed to update the state of the proposal to _"Completed"_ as well. A corresponding issue is closed at this point.

Reviewer: ```suggestion
8. The proposal state is changed to "Completed"_ and any corresponding issues are closed once the implementation is compete, and the code is merged into the master branch.
```

### Note

> * Committers and PMCs are supposed to label issues with an appropriate label to track its status.

Reviewer: ```suggestion
* Committers and the PMC are supposed to label issues with an appropriate label to track issue and pull request status.
```

> * There are 4 labels(`draft`, `in-progress`, `completed`, and `abandoned`) to specify the state of a proposal and one special label(`proposal`) to differentiate proposals from other issues.

Reviewer: ```suggestion
* There are 4 labels(`draft`, `in-progress`, `completed`, and `abandoned`) to specify the state of a proposal and one special label (`proposal`) to differentiate proposals from other issues.
```

### Proposal Lifecycle

> A proposal is supposed to be in one of the following states:

Reviewer: ```suggestion
A proposal may be in one of the following states:
```

* **Draft**: A new enhancement is proposed and it is under discussion.
* **In-progress**: A consensus for the proposal is formed and implementation is in progress.
* **Completed**: Implementation is finished and the change is included in the master branch.
* **Abandoned**: A proposal is not accepted for some reason such as ”no consensus is formed”.

## Issue

> Abandoned proposals would also be filed in this directory for history.

Reviewer: ```suggestion
Abandoned proposals are filed in this directory for archival.
```

<!--
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
-->

# Title

> Proposal for OpenWhisk EnhanceMents (POEMs)

Reviewer: ```suggestion
Process for introducing an OpenWhisk Enhancement (POEM)
```

## Status
* Current state: Draft
* Author: @style95

## Summary

> A Proposal for OpenWhisk EnhanceMent(POEM) is a way to propose and manage new features/ideas in OpenWhisk.

Reviewer: ```suggestion
Introduce a process for suggesting and implementing a substantive OpenWhisk Enhancement (POEM).
```

> The one who wants to include any major new feature, subsystem, piece of functionality and any breaking changes is supposed to open a proposal under this directory.

Reviewer: ```suggestion
A developer or group of developers working together to propose and implement a major new feature or functionality, a new subsystem, or a breaking change should follow the process described herein and open a proposal for consideration under this directory.
```
